# Introduction

This script contains a workflow that calculates the pairwise dNdS ratio of a single 1-to-1 orthogroup created with Orthofinder wit PAML.

### Quick-start
Also use `-h` for more info
```
python3 calculate_orthogroup_dNdS.py --orthogroup /path/to/orthofinder/Orthogroup_Sequences/OG0006980.fa --cds /path/to/dir/cds_complete --pal2nalbin /path/to/pal2nal.pl --verbose --overwrite
```

## Dependencies

* [Orthofinder](https://github.com/davidemms/OrthoFinder) (The script assumes the file structure of orthofinder output to create the default output directory and to retrieve the gene tree)
* Python3 (libraries: os, subprocess, re, time)
* [BioPython](https://biopython.org/wiki/Download) (specifically SeqIO and AlignIO)
* [Clustal Omega](http://www.clustal.org/omega/) (for the alignment)
* [PAML](http://abacus.gene.ucl.ac.uk/software/paml.html) (the script has yn00 and codeml implemented)
* [pal2nal](https://www.bork.embl.de/pal2nal/) (to make the protein alignment codon-based)
  (Here I allow to give a custom path to the executeable since it wasn't installed in the HPC already)

## Input data

* `--orthogroup`: The orthogroup protein sequence file of a 1-to-1 ortholog generated by orthofinder (in the Orthogroup_Sequences directory)
* `--cds`: nucleotide sequences (for the codon-based alignment), and there are two options here:
    * _Directory_: contains fasta files of all cds (nucleotide sequences!) for each species in your orthogroup. The script then makes a nucleotide fasta file corresponding to the input orthogroup by extracting all cds sequences with matching headers. Therefore, the headers have to match the protein-fasta files from the orthofinder run!
    * _File_: A nucleotide fasta file containing the nucleotide sequences that correspond to the protein sequences in the orthogroup input

## Workflow

1. 
